### ISSUE TITLE
Accessibility: Signaling Model

### ISSUE THREAD

**Issue by carlos-zamora** [microsoft/terminal#2447]
# Description of the new feature/enhancement

UIA has signal events ([link](https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-event-ids)). ConHost supported them. WindowUiaProvider was an integral part to that.

Windows Terminal currently doesn't support them. We have some architecture in place to do so though.

# Proposed technical implementation details (optional)

WindowUiaProvider already has the groundwork for that. It might be that we don't even need a WindowUiaProvider and just have XAML do some of it.

For reference, the signaling model for conhost was as follows:

<ConHost Event (i.e.: new text)>
--> Signal WindowUiaProvider
--> Signal ScreenInfoUiaProvider
--> Raise UiaAutomationEvent to Client

The client then responds to these events with more actions (i.e.: recreate object, etc...)

EDIT: The functionality for signaling includes...
- [x] Selection Events (selection changed) [PR](https://github.com/microsoft/terminal/pull/2989)
- [ ] Text Buffer Events (new output)
- [ ] Scroll Events

I may possible need...
- [ ] Resize events
- [ ] Cursor Events

**Comment by codeofdusk** [comment]
Blocking nvaccess/nvda#10305.

**Comment by carlos-zamora** [comment]
@codeofdusk @michaelDCurran @leonardder
CC @DHowett-MSFT 

I've attached a release x64 build of `dev/cazamor/acc/sig/output`. I'll be working off of that branch to get automation events working for output. The changes are pretty naive but simple: whenever we need to render new content to the screen, fire a TextChanged automation event.

For something as simple as `echo hello`, I can see that some degenerate ranges are created at the end of the written command, then at the end of the prompt after the command has been executed. However, it seems like a range connecting both points is never actually created and read.

Are you expecting a different Automation Event? Do I need to configure my TermControlAutomationPeer any further, maybe? Let me know what's going on from your end :)

NOTE: you'll have to run WindowsTerminal.exe here, NOT OpenConsole.exe. ConHost has not been switched over to the new signaling model yet, whereas Windows Terminal has.

[WT_signal_output.zip](https://github.com/microsoft/terminal/files/4299459/WT_signal_output.zip)

**Comment by codeofdusk** [comment]
> @codeofdusk @michaelDCurran @leonardder
> CC @DHowett-MSFT
> 
> I've attached a release x64 build of `dev/cazamor/acc/sig/output`. I'll be working off of that branch to get automation events working for output. The changes are pretty naive but simple: whenever we need to render new content to the screen, fire a TextChanged automation event.
> 
> For something as simple as `echo hello`, I can see that some degenerate ranges are created at the end of the written command, then at the end of the prompt after the command has been executed. However, it seems like a range connecting both points is never actually created and read.
> 
> Are you expecting a different Automation Event? Do I need to configure my TermControlAutomationPeer any further, maybe? Let me know what's going on from your end :)
> 
> NOTE: you'll have to run WindowsTerminal.exe here, NOT OpenConsole.exe. ConHost has not been switched over to the new signaling model yet, whereas Windows Terminal has.
> 
> [WT_signal_output.zip](https://github.com/microsoft/terminal/files/4299459/WT_signal_output.zip)

Thanks @carlos-zamora This looks sufficient for implementing terminal support in NVDA. Things are mostly working (see nvaccess/nvda#10784), except we're getting doubled characters when typing (but I think I know where the issue is, and it's on our side). Are there plans to have this merged soon?

**Comment by carlos-zamora** [comment]
> > @codeofdusk @michaelDCurran @leonardder
> > CC @DHowett-MSFT
> > I've attached a release x64 build of `dev/cazamor/acc/sig/output`. I'll be working off of that branch to get automation events working for output. The changes are pretty naive but simple: whenever we need to render new content to the screen, fire a TextChanged automation event.
> > For something as simple as `echo hello`, I can see that some degenerate ranges are created at the end of the written command, then at the end of the prompt after the command has been executed. However, it seems like a range connecting both points is never actually created and read.
> > Are you expecting a different Automation Event? Do I need to configure my TermControlAutomationPeer any further, maybe? Let me know what's going on from your end :)
> > NOTE: you'll have to run WindowsTerminal.exe here, NOT OpenConsole.exe. ConHost has not been switched over to the new signaling model yet, whereas Windows Terminal has.
> > [WT_signal_output.zip](https://github.com/microsoft/terminal/files/4299459/WT_signal_output.zip)
> 
> Thanks @carlos-zamora This looks sufficient for implementing terminal support in NVDA. Things are mostly working (see [nvaccess/nvda#10784](https://github.com/nvaccess/nvda/pull/10784)), except we're getting doubled characters when typing (but I think I know where the issue is, and it's on our side). Are there plans to have this merged soon?

If I can get it working for Narrator too, that'd be ideal. But, I guess there's no harm in merging what I have then iterating on master. I'll submit a PR today then. Thanks!

**Comment by codeofdusk** [comment]
> > > @codeofdusk @michaelDCurran @leonardder
> > > CC @DHowett-MSFT
> > > I've attached a release x64 build of `dev/cazamor/acc/sig/output`. I'll be working off of that branch to get automation events working for output. The changes are pretty naive but simple: whenever we need to render new content to the screen, fire a TextChanged automation event.
> > > For something as simple as `echo hello`, I can see that some degenerate ranges are created at the end of the written command, then at the end of the prompt after the command has been executed. However, it seems like a range connecting both points is never actually created and read.
> > > Are you expecting a different Automation Event? Do I need to configure my TermControlAutomationPeer any further, maybe? Let me know what's going on from your end :)
> > > NOTE: you'll have to run WindowsTerminal.exe here, NOT OpenConsole.exe. ConHost has not been switched over to the new signaling model yet, whereas Windows Terminal has.
> > > [WT_signal_output.zip](https://github.com/microsoft/terminal/files/4299459/WT_signal_output.zip)
> > 
> > 
> > Thanks @carlos-zamora This looks sufficient for implementing terminal support in NVDA. Things are mostly working (see [nvaccess/nvda#10784](https://github.com/nvaccess/nvda/pull/10784)), except we're getting doubled characters when typing (but I think I know where the issue is, and it's on our side). Are there plans to have this merged soon?
> 
> If I can get it working for Narrator too, that'd be ideal. But, I guess there's no harm in merging what I have then iterating on master. I'll submit a PR today then. Thanks!

I've tested with Narrator and observed, as you did, that we aren't getting new text announcements.

In NVDA, we have an `NVDAObjects.behaviors.LiveText` class that waits for text change events from various APIs (such as UIA), diffs the current text against the last-reported text, and speaks the changes. Does Narrator have a similar paraddigm for consoles?

**Comment by carlos-zamora** [comment]
> > > > @codeofdusk @michaelDCurran @leonardder
> > > > CC @DHowett-MSFT
> > > > I've attached a release x64 build of `dev/cazamor/acc/sig/output`. I'll be working off of that branch to get automation events working for output. The changes are pretty naive but simple: whenever we need to render new content to the screen, fire a TextChanged automation event.
> > > > For something as simple as `echo hello`, I can see that some degenerate ranges are created at the end of the written command, then at the end of the prompt after the command has been executed. However, it seems like a range connecting both points is never actually created and read.
> > > > Are you expecting a different Automation Event? Do I need to configure my TermControlAutomationPeer any further, maybe? Let me know what's going on from your end :)
> > > > NOTE: you'll have to run WindowsTerminal.exe here, NOT OpenConsole.exe. ConHost has not been switched over to the new signaling model yet, whereas Windows Terminal has.
> > > > [WT_signal_output.zip](https://github.com/microsoft/terminal/files/4299459/WT_signal_output.zip)
> > > 
> > > 
> > > Thanks @carlos-zamora This looks sufficient for implementing terminal support in NVDA. Things are mostly working (see [nvaccess/nvda#10784](https://github.com/nvaccess/nvda/pull/10784)), except we're getting doubled characters when typing (but I think I know where the issue is, and it's on our side). Are there plans to have this merged soon?
> > 
> > 
> > If I can get it working for Narrator too, that'd be ideal. But, I guess there's no harm in merging what I have then iterating on master. I'll submit a PR today then. Thanks!
> 
> I've tested with Narrator and observed, as you did, that we aren't getting new text announcements.
> 
> In NVDA, we have an `NVDAObjects.behaviors.LiveText` class that waits for text change events from various APIs (such as UIA), diffs the current text against the last-reported text, and speaks the changes. Does Narrator have a similar paraddigm for consoles?

Not sure. Their code can be a bit complicated haha. I'm working with one of their devs to get a better look. I'm hoping that it can get fixed by next week because we're definitely VERY close.

**Comment by leonardder** [comment]
If narrator doesn't use a diffing method like NVDA does, the only alternative I can think of is making the terminal a live region. Pretty sure it doesn't work like this, though, as it has several drawbacks.

**Comment by leonardder** [comment]
Ah and in theory, UIA notification events could be used as well. If they are fired by conhost, NVDA would ignore them as NVDA only speaks notification events fire by the foreground application.

**Comment by codeofdusk** [comment]
> But, I guess there's no harm in merging what I have then iterating on master.

@carlos-zamora Do you still plan to do this?

**Comment by carlos-zamora** [comment]
> > But, I guess there's no harm in merging what I have then iterating on master.
> 
> @carlos-zamora Do you still plan to do this?

Yup! Here it is: https://github.com/microsoft/terminal/pull/4826
